src/volume_mgr.rs

```diff
@@ -644,32 +733,54 @@ where
                 .write(&blocks, block_idx)
                 .map_err(Error::DeviceError)?;
             written += to_copy;
-            file.current_cluster = current_cluster;
-            let to_copy = i32::try_from(to_copy).map_err(|_| Error::ConversionError)?;
-            // TODO: Should we do this once when the whole file is written?
-            file.update_length(file.length + (to_copy as u32));
-            file.seek_from_current(to_copy).unwrap();
-            file.entry.attributes.set_archive(true);
-            file.entry.mtime = self.timesource.get_timestamp();
-            debug!("Updating FAT info sector");
-            match &mut volume.volume_type {
-                VolumeType::Fat(fat) => {
-                    fat.update_info_sector(self)?;
-                    debug!("Updating dir entry");
-                    self.write_entry_to_disk(fat.get_fat_type(), &file.entry)?;
-                }
+            self.open_files[file_idx].current_cluster = current_cluster;
+
+            let to_copy = to_copy as u32;
+            let new_offset = self.open_files[file_idx].current_offset + to_copy;
+            if new_offset > self.open_files[file_idx].entry.size {
+                // We made it longer
+                self.open_files[file_idx].update_length(new_offset);
             }
+            self.open_files[file_idx]
+                .seek_from_start(new_offset)
+                .unwrap();
+            // Entry update deferred to file close, for performance.
         }
+        self.open_files[file_idx].entry.attributes.set_archive(true);
+        self.open_files[file_idx].entry.mtime = self.time_source.get_timestamp();
         Ok(written)
     }
 
     /// Close a file with the given full path.
-    pub fn close_file(&mut self, volume: &Volume, file: File) -> Result<(), Error<D::Error>> {
-        // Unwrap, because we should never be in a situation where we're attempting to close a file
-        // with an ID which doesn't exist in our open files list.
-        let idx_to_close = cluster_position_by_id(&self.open_files, file.search_id).unwrap();
-        self.open_files.remove(idx_to_close);
+    pub fn close_file(&mut self, file: File) -> Result<(), Error<D::Error>> {
+        let mut found_idx = None;
+        for (idx, info) in self.open_files.iter().enumerate() {
+            if file == info.file_id {
+                found_idx = Some((info, idx));
+                break;
+            }
+        }
+
+        let (file_info, file_idx) = found_idx.ok_or(Error::BadHandle)?;
+
+        if file_info.dirty {
+            let volume_idx = self.get_volume_by_id(file_info.volume_id)?;
+            match self.open_volumes[volume_idx].volume_type {
+                VolumeType::Fat(ref mut fat) => {
+                    debug!("Updating FAT info sector");
+                    fat.update_info_sector(&self.block_device)?;
```

Comment:
I wonder if we should do this more often than just on file close to alleviate issues with crashing firmware. Is there a periodic task where things get flushed to disk?

Reply:
Maybe just adding an API to do it and tell the user to periodically handle it. _or_ check during file IO if too much time has passed since the last flush and just update things to make sure they stay "sort of" in sync? Does this help prevent corruption at all?

Reply:
A flush API could work. It's equivalent to a close and reopen without losing your seek position.

Reply:
That's #98